src/SenseiConfig.hs

```diff
@@ -0,0 +1,82 @@
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE RecordWildCards #-}
+
+module SenseiConfig where
+
+import GHC.Generics
+import Data.Aeson
+import Data.Yaml
+import System.Directory
+import System.Process
+import System.Exit
+import Text.Casing
+
+senseiGlobalConfigFilename :: FilePath
+senseiGlobalConfigFilename = "senseirc"
+
+senseiLocalConfigFilename :: FilePath
+senseiLocalConfigFilename = ".senseirc"
+
+data SenseiConfig = SenseiConfig {
+  -- | Shell command to run after a file has changed, but before reloading. Cancels the current
+  -- trigger cycle when exiting with non-zero exit code.
+  senseiConfigBeforeReload :: Maybe String
+  -- | Shell command to run after a successful reload, but before running the tests. Cancels the
+  -- current trigger cycle when exiting with non-zero exit code.
+, senseiConfigAfterSuccessfulReload :: Maybe String
+  -- | Shell command to run after the trigger cycle completed with failure. Has access to the result
+  -- via seito.
+, senseiConfigOnFailure :: Maybe String
+  -- | Shell command to run after the trigger cycle successfully completed. Has access to the result
+  -- via seito.
+, senseiConfigOnSuccess :: Maybe String
+} deriving (Generic, Show, Eq)
+
+instance FromJSON SenseiConfig where
+  parseJSON = genericParseJSON defaultOptions {
+      fieldLabelModifier = toCamel . dropPrefix . dropPrefix . fromAny
+    }
+
+data SenseiHooks = SenseiHooks {
+  senseiHooksBeforeReload :: IO ExitCode
+, senseiHooksAfterSuccessfulReload :: IO ExitCode
+, senseiHooksOnSuccess :: IO ExitCode
+, senseiHooksOnFailure :: IO ExitCode
+}
+
+noop :: IO ExitCode
+noop = return ExitSuccess
+
+noopHooks :: SenseiHooks
+noopHooks = SenseiHooks {
+    senseiHooksBeforeReload = noop
+  , senseiHooksAfterSuccessfulReload = noop
+  , senseiHooksOnSuccess = noop
+  , senseiHooksOnFailure = noop
+  }
+
+-- | Try to load SenseiConfig. Returns Nothing when the file doesn't exist, throws exceptions on
+-- parse errors.
+loadSenseiConfig :: FilePath -> IO (Maybe SenseiConfig)
+loadSenseiConfig filename = do
+  doesFileExist filename >>= \ case
+    True -> decodeFileThrow filename
+    False -> return Nothing
+
+mkSenseiHooks :: Maybe SenseiConfig -> SenseiHooks
+mkSenseiHooks mSenseiConfig = case mSenseiConfig of
+  Nothing -> noopHooks
+  Just (SenseiConfig {..}) -> SenseiHooks {
+      senseiHooksBeforeReload = callMbCommand senseiConfigBeforeReload
+    , senseiHooksAfterSuccessfulReload = callMbCommand senseiConfigAfterSuccessfulReload
+    , senseiHooksOnSuccess = callMbCommand senseiConfigOnSuccess
+    , senseiHooksOnFailure = callMbCommand senseiConfigOnFailure
+    }
+  where
+  callMbCommand :: Maybe String -> IO ExitCode
+  callMbCommand Nothing = noop
+  callMbCommand (Just cmd) = callCommandExitCode cmd
+
+callCommandExitCode :: String -> IO ExitCode
+callCommandExitCode cmd = spawnCommand cmd >>= waitForProcess
```

Comment:
I think this is actually just `system` (it's in the "deprecated" section, but as long as it's not really deprecated, I would not have any issue using it).